WordPress/gutenberg
# Components: Introduce portal-based slot / fill alternative, render Popover as slot / fill

This pull request seeks to refactor Popover to simplify rendering as a slot / fill pair, rather than require a custom context provider providing a DOM node render target.

__Implementation notes:__

Originally I sought to implement this using `react-slot-fill` as we use elsewhere, but quickly encountered similar issues as discovered in #2734 where lifecycle of slot content is not always predictable since component updates are [reflected only after a subsequent state change](https://github.com/camwest/react-slot-fill/blob/f4087870aa3f74b86cb55f882e9f398586800256/src/lib/components/Slot.ts#L53). 

Recognizing that React 16's portals feature provides the tools for a Slot / Fill implementation, I sought to explore what would be necessary for a first-party implementation. Happily it seemed to work quite well, needing only a context provider to register, unregister, and retrieve `ref` nodes of the rendered `Slot`s to which a `Fill` can create a portal. The first-party portal implementation does not suffer from the same lifecycle issues as `react-slot-fill`. Further encouraging is that such an approach [appears advocated by Dominic Gannaway](https://wordpress.slack.com/archives/C5UNMSU4R/p1506440954000270), a React core team member.

We should decide whether this is worth using as a substitute to `react-slot-fill`. It may be worthwhile to split this pull request into two, where the first would convert existing usage to the new WordPress-specific `slot-fill`.

__Testing instructions:__

Verify that there are no regressions in Popover usage, particularly:

- Different types of popovers (tooltips, dropdown)
- Focus should transition into popover when opened (e.g. Inserter)
- Popover should reorient itself when lacking sufficient space to open in the desired direction
- Popover should close when clicking outside
- No console errors should be logged in any of the above
